packages/client/src/client.ts

```diff
@@ -0,0 +1,165 @@
+
+  /** Returns a new or re-used `PromiseClient<T>` for a specific `PenumbraService` */
+  readonly service: <T extends PenumbraService>(
+    service: T,
+    options: Omit<ChannelTransportOptions, 'getPort'>,
+  ) => PromiseClient<T>;
+}
+
+export class PenumbraClient implements IPenumbraClient {
+  private readonly callbacks = new Set<(detail: PenumbraStateEventDetail) => void>();
+
+  private origin?: string;
+  private provider?: PenumbraProvider;
+  private port?: MessagePort;
+  private serviceClients: Map<string, PromiseClient<PenumbraService>>;
+
+  constructor() {
+    this.serviceClients = new Map();
+  }
+
+  private assertPort() {
+    if (!this.port) {
+      throw new PenumbraProviderNotConnectedError(this.origin);
+    }
+    return this.port;
+  }
+
+  private assertProvider() {
+    if (!this.provider) {
+      throw new PenumbraProviderNotConnectedError(this.origin);
+    }
+    return this.provider;
+  }
+
+  public async connect(requireOrigin?: string): Promise<string> {
+    const providerOrigin = requireOrigin ?? availableOrigin();
+
+    await assertProviderManifest(providerOrigin);
+
+    this.origin = providerOrigin;
+    this.provider = assertProviderRecord(providerOrigin);
+
+    const request = this.provider.connect().then(port => {
+      this.port = port;
+    });
+
+    this.provider.addEventListener('penumbrastate', (evt: PenumbraStateEvent | Event) => {
+      if (isPenumbraStateEvent(evt)) {
+        void request.finally(() => this.callbacks.forEach(cb => cb(evt.detail)));
+      }
+    });
+
+    await request;
+
+    return this.provider.manifest;
+  }
+
+  public async reconnect(requireOrigin?: string) {
+    const providers = assertPenumbra();
+
+    let provider: PenumbraProvider | undefined;
+    let origin = requireOrigin;
+    if (requireOrigin) {
+      provider = assertProviderRecord(requireOrigin);
+    } else {
+      const connectedEntry = Object.entries(providers).find(([, provider]) =>
+        provider.isConnected(),
+      );
+      origin = connectedEntry?.[0];
+      provider = connectedEntry?.[1];
+    }
+
+    if (!origin || !provider?.isConnected()) {
+      throw new PenumbraProviderNotConnectedError(origin);
+    }
+
+    await assertProviderManifest(origin);
+
+    this.origin = origin;
+    this.provider = provider;
+    this.port = await provider.connect();
+    this.callbacks.forEach(cb => cb(new PenumbraStateEvent(origin, provider.state(), true).detail));
+
+    return origin;
+  }
+
+  public disconnect() {
+    const request = this.assertProvider().disconnect();
+    this.assertPort().close();
+    this.serviceClients.clear();
+    return request;
+  }
+
+  public isConnected() {
+    return this.provider?.isConnected() ?? false;
+  }
+
+  public getState() {
+    return this.assertProvider().state();
+  }
+
+  public onConnectionChange(callback: (detail: PenumbraStateEventDetail) => void) {
+    this.callbacks.add(callback);
+  }
```

Comment:
it seems like there's now no way to remove a callback

Reply:
copied the AbortController usage here